### PR TITLE
Use containerd binares from kind-ci

### LINKED ISSUE #870: Kind can't pull Docker images from Github's pkg registry
I am using kind `v0.5.1` to pull Docker image from Github's new pkg registry. But this is failing. The exact same commands are working in minikube `v1.4.0`.

```
export DOCKER_TOKEN=**my-personal-token-with-pkg-read-write-permission**
export USERNAME=stashed
kubectl create secret docker-registry regcred --namespace=kube-system --docker-server=docker.pkg.github.com --docker-username=${USERNAME} --docker-password=${DOCKER_TOKEN}
```

Now, I create the following pod.
```
apiVersion: v1
kind: Pod
metadata:
  name: private-reg
  namespace: kube-system
spec:
  restartPolicy: Never
  containers:
  - name: private-reg-container
    image: docker.pkg.github.com/stashed/stash/stash:v0.9.0-rc.0-25-g1aa27c95_linux_amd64
  imagePullSecrets:
  - name: regcred
```

If I describe the pod, I see the following error:

```
Events:
  Type     Reason     Age                  From                         Message
  ----     ------     ----                 ----                         -------
  Normal   Scheduled  2m27s                default-scheduler            Successfully assigned kube-system/private-reg to kind-control-plane
  Normal   Pulling    62s (x4 over 2m25s)  kubelet, kind-control-plane  Pulling image "docker.pkg.github.com/stashed/stash/stash:v0.9.0-rc.0-25-g1aa27c95_linux_amd64"
  Warning  Failed     62s (x4 over 2m24s)  kubelet, kind-control-plane  Failed to pull image "docker.pkg.github.com/stashed/stash/stash:v0.9.0-rc.0-25-g1aa27c95_linux_amd64": rpc error: code = Unknown desc = failed to resolve image "docker.pkg.github.com/stashed/stash/stash:v0.9.0-rc.0-25-g1aa27c95_linux_amd64": no available registry endpoint: unexpected status code https://docker.pkg.github.com/v2/stashed/stash/stash/manifests/v0.9.0-rc.0-25-g1aa27c95_linux_amd64: 406 Not Acceptable
  Warning  Failed     62s (x4 over 2m24s)  kubelet, kind-control-plane  Error: ErrImagePull
  Warning  Failed     49s (x6 over 2m24s)  kubelet, kind-control-plane  Error: ImagePullBackOff
  Normal   BackOff    38s (x7 over 2m24s)  kubelet, kind-control-plane  Back-off pulling image "docker.pkg.github.com/stashed/stash/stash:v0.9.0-rc.0-25-g1aa27c95_linux_amd64"

```

I followed the steps in https://github.com/kubernetes-sigs/kind/issues/817 and I get the same error:

```
root@kind-control-plane:/# crictl pull docker.pkg.github.com/stashed/stash/stash:v0.9.0-rc.0-25-g1aa27c95_linux_amd64
FATA[0000] pulling image failed: rpc error: code = Unknown desc = failed to resolve image "docker.pkg.github.com/stashed/stash/stash:v0.9.0-rc.0-25-g1aa27c95_linux_amd64": no available registry endpoint: unexpected status code https://docker.pkg.github.com/v2/stashed/stash/stash/manifests/v0.9.0-rc.0-25-g1aa27c95_linux_amd64: 406 Not Acceptable 
```

So, the error is `406 Not Acceptable` . I am able to pull this image on my desktop's docker daemon. So, this seems like a kind specific issue.

### PR DESCRIPTION
**Pull request by aojea**
Containerd releases are not available in distributions, also it is common that we need some features that are recently merged in containerd, i.e. dual-stack, github registry support, ....

This situation make ideal for us working with nightlies, and we submitted a patch https://github.com/containerd/containerd/pull/3702 to the project to make them available from their repo.

In the meantime, I've created a repo that's publishing nightlies of containerd.

This also allows us to get rid of the containerd systemd overrides and use our own systemd init file.

Fixes #870

### CODE REVIEW

**Comment by aojea** [comment]
/assign @BenTheElder 

just a placeholder, maybe we can create a kind-containerd-nigthly under https://github.com/kind-ci ?

File: images/base/Dockerfile [hunk 1]
```diff
@@ -69,18 +68,27 @@ RUN clean-install \
...
+# Install containerd nigthly binaries so we can use latest features not present in Ubuntu
+# https://github.com/aojea/containerd-nightly-releases/releases/download/v0.0.0-20191003172204-be6bead4/containerd-0.0.0-20191003172204-be6bead4.linux-amd64.tar.gz
+ARG CONTAINERD_VERSION="0.0.0-20191003172204-be6bead4"
+ARG CONTAINERD_BASE_URL="https://github.com/aojea/containerd-nightly-releases/releases/download/"
+RUN export ARCH=$(dpkg --print-architecture | sed 's/ppc64el/ppc64le/' | sed 's/armhf/arm/') \
+    && export CONTAINERD_TARBALL="v${CONTAINERD_VERSION}/containerd-${CONTAINERD_VERSION}.linux-${ARCH}.tar.gz" \
+    && echo "${CONTAINERD_URL}" \
+    && export CONTAINERD_URL="${CONTAINERD_BASE_URL}${CONTAINERD_TARBALL}" \
+    && curl -sSL --retry 5 --output /tmp/containerd.tgz "${CONTAINERD_URL}" \
+    && sha256sum /tmp/containerd.tgz \
```
Comment (tao12345666333) [inline images/base/Dockerfile]: It’s not necessary？ We don’t check it

**Comment by tao12345666333** [comment]
SGTM 

But should we consider more?  If we use the nightly version of containerd then the cluster created with kind will no longer be a stable cluster.

File: images/base/Dockerfile (hunk 1 above)
Comment (aojea) [reply images/base/Dockerfile]: copy and paste leftover from the cni one

File: images/base/containerd.service [hunk 2]
```diff
@@ -0,0 +1,27 @@
+[Unit]
```
Comment (BenTheElder) [inline images/base/containerd.service]: is this file not in the tarballs? (that's fine, just slightly surprised)

File: images/base/Dockerfile [hunk 3]
```diff
@@ -69,18 +68,26 @@ RUN clean-install \
...
-    && chmod +x /usr/local/bin/ctr \
     && echo "done installing packages"
 
 # override the rp_filter settings to enable calico cni to "just work"
 COPY 10-network-security.conf /etc/sysctl.d/
 
-# add overrides to containerd
-COPY 10-limits.conf 10-restart.conf /etc/systemd/system/containerd.service.d/
+# Install containerd nigthly binaries so we can use latest features not present in Ubuntu
+# https://github.com/aojea/containerd-nightly-releases/releases/download/v0.0.0-20191003172204-be6bead4/containerd-0.0.0-20191003172204-be6bead4.linux-amd64.tar.gz
```
Comment (BenTheElder) [inline images/base/Dockerfile]: stale comment

File: images/base/containerd.service (hunk 2 above)
Comment (aojea) [reply images/base/containerd.service]: it is not, there is only binaries

Comment (BenTheElder) [reply images/base/containerd.service]: can we add a comment about the source with a link to the upstream file? for cross referencing

**Comment by BenTheElder** [comment]
> But should we consider more? If we use the nightly version of containerd then the cluster created with kind will no longer be a stable cluster.

this is a good point, we discussed out of band (while discussing logistics of hosting the nightlies) -- this PR now does `1.3.0` (an actual release-tagged version) via the same system as the nightlies.

the nightlies will also be useful to experiment with, and this now makes it a bit easier to `cd image/base; docker build -arg CONTAINERD_VERSION=foo ...` and test out out a nighty.

we'll revisit wether we actually want to use non-release versions in the future, for now this lets us use versions well ahead of ubuntu's packaging

**Comment by aojea** [comment]
/hold

I have to test it well, I did a quick test and the create cluster timed out,

**Comment by aojea** [comment]
it's failing because of this

> rol-plane kubelet[238]: E1005 06:47:32.007355     238 kuberuntime_sandbox.go:68] CreatePodSandbox for pod "kube-controller-manager-kin
d-control-plane_kube-system(1690996e5a610f4223232aa4b8920c52)" failed: rpc error: code = Unknown desc = failed to create containerd task: OCI runtime create fa
iled: unable to retrieve OCI runtime error (open /run/containerd/io.containerd.runtime.v2.task/k8s.io/ce21230912eb563b93b3141a9c41fdaabdf1734ecef9db9fc308850a5
fc8023c/log.json: no such file or directory): exec: "runc": executable file not found in $PATH: unknown

**Comment by aojea** [comment]
/hold cancel

it turns out containerd requires a specific version of runc, https://github.com/containerd/containerd/blob/master/RUNC.md

It's working for me locally, with stable and nigthlies, remember that we can pass the containerd version as argument when building the base image :-)

**Comment by neolit123** [comment]
@aojea 
odd, this makes containerd not exactly easy to setup without package managers.
do you happen to know if distros a handling this dependency properly in their packages?

**Comment by aojea** [comment]
> @aojea
> odd, this makes containerd not exactly easy to setup without package managers.
> do you happen to know if distros are handling this dependency properly in their packages?

it's ok, the `runc` sync is required for development, but they seem to synchronize with runc releases when the release stable versions, it's explained in the link attached 

> After an official runc release we will start pinning containerd support to a specific version but various development and testing features may require a newer runc version than the latest release. If you encounter any runtime errors, please make sure your runc is in sync with the commit/tag provided in this document

The distros are handling it ok, they require the specific `runc`version

Ubuntu https://launchpad.net/ubuntu/+source/containerd/1.2.9-0ubuntu1
OpenSUSE https://build.opensuse.org/package/view_file/Virtualization:containers/containerd/containerd.spec?expand=1

**Comment by BenTheElder** [comment]
> odd, this makes containerd not exactly easy to setup without package managers.
> do you happen to know if distros are handling this dependency properly in their packages?

_all_ container engines are bundling runc as a specific dependency (eg crio has [crio-runc](https://launchpad.net/~projectatomic/+archive/ubuntu/ppa)) which makes installing two of them tricky unless they coordinate. this is expected to change across all of them in the future as runc stabilizes ™️ 

runc is expected to be built with containerd https://github.com/containerd/containerd/blob/master/BUILDING.md#build-containerd-and-runc

**Comment by BenTheElder** [comment]
[squashed commits, pushed and updated to new base image, dropped the `ctr` build we no longer need]

**Comment by BenTheElder** [comment]
prow flake
/retest

**Comment by BenTheElder** [comment]
whoah we skipped all of the tests?? https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/920/pull-kind-e2e-kubernetes/1180600069666115587 !?

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/test-infra/pull/14631

**Comment by BenTheElder** [comment]
too fast it seems...
/test pull-kind-e2e-kubernetes

**Comment by BenTheElder** [comment]
/lgtm
/approve
thanks for working on this
/hold cancel

**Pull request merged**